### PR TITLE
extract surname from OCR and add surname to exception

### PR DESCRIPTION
**Pull request by mdayican**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/BPS-894

### Change description ###

extract surname from OCR and add surname to exception

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 1]
```diff
@@ -98,19 +103,29 @@ public void should_create_exception_record_for_new_application() throws Exceptio
...
+            caseDetails.getId(),
+            caseDetails.getCaseTypeId()
+        );
+        // envelope ID from the JSON resource representing the test message
+        assertThat(caseDetails.getData().get("envelopeId")).isEqualTo(messageEnvelopeId);
         assertThat(caseDetails.getCaseTypeId()).isEqualTo("BULKSCAN_ExceptionRecord");
         assertThat(caseDetails.getJurisdiction()).isEqualTo("BULKSCAN");
 
-        Map<String, String> expectedOcrData = ImmutableMap.of("field1", "value1", "field2", "value2");
+        Map<String, String> expectedOcrData = ImmutableMap.of("field1", "value1",
```
Comment (doncem) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: let's put `"field1", "value1",` to new line for consistency

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java [hunk 2]
```diff
@@ -90,4 +99,37 @@ private String setDisplayCaseReferenceFlag(String caseRef, Classification classi
...
+
+    private String extractSurnameFromOcrData(Envelope envelope) {
+        if (CollectionUtils.isEmpty(envelope.ocrData)) {
+            return null;
+        }
+
+        String surnameOcrFieldName = serviceConfigProvider.getConfig(envelope.container)
+            .getSurnameOcrFieldName(envelope.formType);
+
+        List<String> surnameList = envelope.ocrData.stream().filter(ocrData -> ocrData.name.equals(surnameOcrFieldName))
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: let's put `.stream()` and `.filter()` to new lines so it nicely matches the `.map` the line below

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java [hunk 3]
```diff
@@ -90,4 +99,37 @@ private String setDisplayCaseReferenceFlag(String caseRef, Classification classi
...
+            .map(ocrData -> ocrData.value).collect(Collectors.toList());
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: `.collect()` too

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java [hunk 4]
```diff
@@ -90,4 +99,37 @@ private String setDisplayCaseReferenceFlag(String caseRef, Classification classi
...
+        if (surnameList.size() == 0) {
+            LOGGER.info(
+                "Surname not found in OCR data. Surname Ocr Field Name:{}. Envelope id:{},Case Ref:{},Jurisdiction:{}",
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: personally i would consider this as a debug, but don't take my word for it. see what team thinks of it.

p.s. please put single space after `:` and `,`

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java [hunk 5]
```diff
@@ -90,4 +99,37 @@ private String setDisplayCaseReferenceFlag(String caseRef, Classification classi
...
+                surnameOcrFieldName,
+                envelope.id,
+                envelope.caseRef,
+                envelope.jurisdiction
+            );
+            return null;
+        } else if (surnameList.size() > 1) {
+            LOGGER.info(
+                "Surname found {} times in OCR data."
+                    + "Surname Ocr Field Name:{} Envelope id:{},Case Ref:{},Jurisdiction:{}",
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: same comment as above

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -67,6 +67,9 @@ service-config:
       case-type-ids:
         - Bulk_Scanned
       allow-creating-case-before-payments-are-processed: true
+      form-type-to-surname-ocr-field-mappings:
+        - formType: B123
+          ocrField: field_surname
```
Comment (doncem) [inline src/main/resources/application.yaml]: just a note for QA I guess - need to be aware that changes will be visible in ccd UI for B123 form which has `field_surname` inside OCR data 🙂

File: Jenkinsfile_CNP [hunk 7]
```diff
@@ -26,7 +26,16 @@ def secrets = [
     secret('payments-queue-send-connection-string', 'PAYMENTS_QUEUE_WRITE_CONN_STRING')
   ]
 ]
-
+def prSecrets = [
+  's2s-${env}': [
+    secret('microservicekey-bulk-scan-orchestrator', 'S2S_SECRET')
```
Comment (nitinprabhu) [inline Jenkinsfile_CNP]: you can copy  the secret in bulk scan vault and load it from there ?

File: Jenkinsfile_CNP [hunk 8]
```diff
@@ -26,7 +26,16 @@ def secrets = [
...
+  ],
+  'bulk-scan-${env}': [
+    secret('idam-users-bulkscan-username', 'IDAM_USER_NAME'),
+    secret('idam-users-bulkscan-password', 'IDAM_USER_PASSWORD'),
+    secret('idam-client-secret', 'IDAM_CLIENT_SECRET')
```
Comment (nitinprabhu) [inline Jenkinsfile_CNP]: how does the existing functional tests work and why we need specifically for the  surname functional tests

File: src/main/resources/application.yaml (hunk 6 above)
Comment (nitinprabhu) [reply src/main/resources/application.yaml]: in future this will be extending to multiple fields/forms. So probably we can have jurisdiction at top and list of forms and then each form will have list of fields like surname first name etc

File: Jenkinsfile_CNP (hunk 8 above)
Comment (mdayican) [reply Jenkinsfile_CNP]: functional test were sending to the AAT. that is why there were no logs

File: Jenkinsfile_CNP (hunk 7 above)
Comment (mdayican) [reply Jenkinsfile_CNP]: we need to use osba created queue in PR

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java (hunk 4 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: agree but for now we can keep is info as we will need this log while testing definitions

File: src/main/resources/application.yaml (hunk 6 above)
Comment (mdayican) [reply src/main/resources/application.yaml]: this config was there , lets keep it as it is, latter depending on requirements  we can improve it

File: Jenkinsfile_CNP (hunk 7 above)
Comment (nitinprabhu) [reply Jenkinsfile_CNP]: not sure how osba and s2s are related

File: Jenkinsfile_CNP (hunk 8 above)
Comment (nitinprabhu) [reply Jenkinsfile_CNP]: these are idam configs ? and not connection string ?

File: Jenkinsfile_CNP (hunk 7 above)
Comment (nitinprabhu) [reply Jenkinsfile_CNP]: got you now

File: Jenkinsfile_CNP (hunk 8 above)
Comment (mdayican) [reply Jenkinsfile_CNP]: nothing change for idam it is same as before. just exclude connection secrets for PR

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 9]
```diff
@@ -98,19 +103,30 @@ public void should_create_exception_record_for_new_application() throws Exceptio
...
+        // envelope ID from the JSON resource representing the test message
+        assertThat(caseDetails.getData().get("envelopeId")).isEqualTo(messageEnvelopeId);
         assertThat(caseDetails.getCaseTypeId()).isEqualTo("BULKSCAN_ExceptionRecord");
         assertThat(caseDetails.getJurisdiction()).isEqualTo("BULKSCAN");
 
-        Map<String, String> expectedOcrData = ImmutableMap.of("field1", "value1", "field2", "value2");
+        Map<String, String> expectedOcrData = ImmutableMap.of(
+            "field1", "value1",
+            "field2", "value2",
+            "field_surname","surnameXXXX"
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: missing space

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 10]
```diff
@@ -98,19 +103,30 @@ public void should_create_exception_record_for_new_application() throws Exceptio
             .until(() -> findCasesByPoBox(randomPoBox).size() == 1);
 
         CaseDetails caseDetails = findCasesByPoBox(randomPoBox).get(0);
+
+        LOG.info(
+            "PoBox :{},caseDetails ID:{}, getCaseTypeId:{}",
+            randomPoBox,
+            caseDetails.getId(),
+            caseDetails.getCaseTypeId()
+        );
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: do we need it in tests? Also, missing space after `,`

File: Jenkinsfile_CNP [hunk 11]
```diff
@@ -78,5 +87,11 @@ withPipeline(type, product, component) {
...
   disableLegacyDeployment()
   enableSlackNotifications(channel)
-  loadVaultSecrets(secrets)
+
+  onPR() {
+    loadVaultSecrets(prSecrets)
+  }
+  onNonPR() {
+    loadVaultSecrets(nonPrSecrets)
+  }
```
Comment (rkondratowicz) [inline Jenkinsfile_CNP]: what are those changes related to? Not surname ocr field I assume? can it go to a separate PR with explanation?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 12]
```diff
@@ -98,19 +103,30 @@ public void should_create_exception_record_for_new_application() throws Exceptio
...
+            "field2", "value2",
+            "field_surname","surnameXXXX"
+        );
         assertThat(getOcrData(caseDetails)).isEqualTo(expectedOcrData);
-
         List<String> expectedOcrDataWarnings = Arrays.asList("warning 1", "warning 2");
         assertThat(getOcrDataValidationWarnings(caseDetails)).isEqualTo(expectedOcrDataWarnings);
 
-        // envelope ID from the JSON resource representing the test message
-        assertThat(caseDetails.getData().get("envelopeId")).isEqualTo(messageEnvelopeId);
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: why was this removed?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java [hunk 13]
```diff
@@ -164,7 +180,6 @@ void create_exception_record_for_supplementary_evidence_with_ocr() throws Except
             .isEqualTo("SUPPLEMENTARY_EVIDENCE_WITH_OCR");
         assertThat(getOcrData(exceptionRecord)).isEqualTo(expectedOcrData);
         assertThat(getCaseDataForField(exceptionRecord, "envelopeCaseReference")).isEqualTo(envelopeCaseRef);
-        assertThat(getCaseDataForField(exceptionRecord, "envelopeLegacyCaseReference")).isEmpty();
```
Comment (rkondratowicz) [inline src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: why?

File: src/functionalTest/resources/envelopes/new-envelope-with-evidence.json [hunk 14]
```diff
@@ -5,6 +5,7 @@
   "jurisdiction": "BULKSCAN",
   "container": "bulkscan",
   "classification": "NEW_APPLICATION",
+  "form_type": "B123",
```
Comment (rkondratowicz) [inline src/functionalTest/resources/envelopes/new-envelope-with-evidence.json]: not related?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java [hunk 15]
```diff
@@ -28,6 +31,8 @@
 
     private final String documentManagementUrl;
     private final String contextPath;
+    private final ServiceConfigProvider serviceConfigProvider;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExceptionRecordMapper.class);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: it would be nice to separate this static field from other regular fields. Maybe first line in the class as in all other classes.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java [hunk 16]
```diff
@@ -90,4 +98,41 @@ private String setDisplayCaseReferenceFlag(String caseRef, Classification classi
...
+                "Surname not found in OCR data. "
+                    + "Surname Ocr Field Name: {}. Envelope id: {}, Case Ref: {}, Jurisdiction: {}",
+                surnameOcrFieldName,
+                envelope.id,
+                envelope.caseRef,
+                envelope.jurisdiction
+            );
+            return null;
+        } else if (surnameList.size() > 1) {
+            LOGGER.info(
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: maybe this should be a warning?

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java (hunk 10 above)
Comment (rkondratowicz) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: fixed all problems: `"PoBox : {},caseDetails ID: {}, getCaseTypeId: {}"`

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java [hunk 17]
```diff
@@ -184,6 +202,17 @@ public void mapEnvelope_sets_display_case_reference_fields_to_yes_when_envelope_
         assertThat(exceptionRecord.showEnvelopeLegacyCaseReference).isEqualTo("Yes");
     }
 
+    @Test
+    public void mapEnvelope_surname_should_null_when_no_surname_data_in_ocr() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java]: 'should be'

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java [hunk 18]
```diff
@@ -71,13 +87,15 @@ public void mapEnvelope_maps_all_fields_correctly() {
         assertThat(exceptionRecord.envelopeLegacyCaseReference).isEqualTo(envelope.legacyCaseRef);
         assertThat(exceptionRecord.showEnvelopeCaseReference).isEqualTo("No"); // for "New Application"
         assertThat(exceptionRecord.showEnvelopeLegacyCaseReference).isEqualTo("No"); // for "New Application"
+        assertThat(exceptionRecord.surname).isEqualTo("surname1");
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java]: either pass `"surname1"` to `envelopeWithSurname(2);` or retrieve surname field from `envelope` object.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java (hunk 15 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: how about the one below this?
ALLOWED_CLASSIFICATIONS

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java (hunk 17 above)
Comment (mdayican) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java]: sets

File: src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java (hunk 12 above)
Comment (mdayican) [reply src/functionalTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/ExceptionRecordCreationTest.java]: not removed , moved up

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java [hunk 19]
```diff
@@ -184,6 +202,17 @@ public void mapEnvelope_sets_display_case_reference_fields_to_yes_when_envelope_
...
+    @Test
+    public void mapEnvelope_sets_surname_null_when_no_surname_data_in_ocr() {
+        //given
+        Envelope envelope = envelope("CASE_123", "LEGACY_CASE_123", Classification.SUPPLEMENTARY_EVIDENCE_WITH_OCR);
+        // when
+        ExceptionRecord exceptionRecord = mapper.mapEnvelope(envelope);
+
+        // then
+        assertThat(exceptionRecord.surname).isNull();
+    }
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java]: missing test for duplicated surnames in OCR

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java [hunk 20]
```diff
@@ -90,4 +99,41 @@ private String setDisplayCaseReferenceFlag(String caseRef, Classification classi
...
+            LOGGER.error(
+                "Surname found {} times in OCR data. "
+                    + "Surname Ocr Field Name: {}, Envelope id: {}, Case Ref: {}, Jurisdiction: {}",
+                surnameList.size(),
+                surnameOcrFieldName,
+                envelope.id,
+                envelope.caseRef,
+                envelope.jurisdiction
+            );
+        }
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapper.java]: no test for this case

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java (hunk 19 above)
Comment (mdayican) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/model/ccd/mappers/ExceptionRecordMapperTest.java]: added
